src/persist-client/src/read.rs

```diff
@@ -471,18 +523,29 @@ where
         // Hack: Keep this method `&self` instead of `&mut self` by cloning the
         // cached copy of the state, updating it, and throwing it away
         // afterward.
-        let batches = self.machine.clone().snapshot(&as_of).await?;
+        let mut machine = self.machine.clone();
+        let batches = machine.snapshot(&as_of).await?;
         let batches = batches.into_iter().flat_map(|b| {
             let desc = b.desc.clone();
             b.keys.into_iter().map(move |k| (k, desc.clone()))
         });
-        let mut splits = (0..num_splits.get())
-            .map(|_| SnapshotSplit {
+        let mut splits = Vec::new();
+        for _ in 0..num_splits.get() {
+            // Register this split as a new reader so that it gets its own
+            // seqno capability.
+            let split_reader_id = ReaderId::new();
+            let _read_cap = machine.clone_reader(&split_reader_id).await;
```

Comment:
Eek, that's a lot of additional "reader state", right? Just pointing out that I saw it, I have no better idea as of now.

Reply:
I don't think the size matters so much, pretty much anything is small compared to the metadata about batches. I am also worried about the traffic of them all downgrading but yeah, I don't see anything else to do here either